sql/row: TestRowFetcherMVCCMetadata failed

sql/row.TestRowFetcherMVCCMetadata [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19758151?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19758151?buildTab=artifacts#/) on master @ [f92b30e3626c9f32191555da8d929eddc59ca956](https://github.com/cockroachdb/cockroach/commits/f92b30e3626c9f32191555da8d929eddc59ca956):


```
=== RUN   TestRowFetcherMVCCMetadata
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/9db56821b86748e0fc537f3d515778f0/logTestRowFetcherMVCCMetadata177548470
    test_log_scope.go:76: use -show-logs to present logs inline
    fetcher_mvcc_test.go:157: expected [{[1 a a a] false 1747937531343998314.0000000000} {[2 b b b] false 1747937531343998314.0000000000}] got []
    fetcher_mvcc_test.go:185: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9db56821b86748e0fc537f3d515778f0/logTestRowFetcherMVCCMetadata177548470
--- FAIL: TestRowFetcherMVCCMetadata (0.71s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowFetcherMVCCMetadata.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50912

--- a/pkg/sql/row/BUILD.bazel
+++ b/pkg/sql/row/BUILD.bazel
@@ -127,6 +127,7 @@ go_test(
         "//pkg/sql",
         "//pkg/sql/catalog",
         "//pkg/sql/catalog/bootstrap",
+        "//pkg/sql/catalog/colinfo",
         "//pkg/sql/catalog/descpb",
         "//pkg/sql/catalog/descs",
         "//pkg/sql/catalog/desctestutils",

--- a/pkg/sql/row/fetcher.go
+++ b/pkg/sql/row/fetcher.go
@@ -1284,12 +1284,6 @@ func (rf *Fetcher) NextRowDecodedInto(
 	return true, spanID, nil
 }
 
-// RowLastModified may only be called after NextRow has returned a non-nil row
-// and returns the timestamp of the last modification to that row.
-func (rf *Fetcher) RowLastModified() hlc.Timestamp {
-	return rf.table.rowLastModified
-}
-
 // RowIsDeleted may only be called after NextRow has returned a non-nil row and
 // returns true if that row was most recently deleted. This method is only
 // meaningful when the configured KVBatchFetcher returns deletion tombstones, which
@@ -1306,7 +1300,7 @@ func (rf *Fetcher) finalizeRow() error {
 		// TODO (rohany): Datums are immutable, so we can't store a DDecimal on the
 		//  fetcher and change its contents with each row. If that assumption gets
 		//  lifted, then we can avoid an allocation of a new decimal datum here.
-		dec := rf.args.Alloc.NewDDecimal(tree.DDecimal{Decimal: eval.TimestampToDecimal(rf.RowLastModified())})
+		dec := rf.args.Alloc.NewDDecimal(tree.DDecimal{Decimal: eval.TimestampToDecimal(rf.table.rowLastModified)})
 		table.row[table.timestampOutputIdx] = rowenc.EncDatum{Datum: dec}
 	}
 	if table.oidOutputIdx != noOutputColumn {

--- a/pkg/sql/row/fetcher_mvcc_test.go
+++ b/pkg/sql/row/fetcher_mvcc_test.go
@@ -15,11 +15,11 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/bootstrap"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/desctestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/fetchpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/row"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
-	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
@@ -74,7 +74,17 @@ func TestRowFetcherMVCCMetadata(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
-	srv, db, kvDB := serverutils.StartServer(t, base.TestServerArgs{})
+	srv, db, kvDB := serverutils.StartServer(t, base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			Store: &kvserver.StoreTestingKnobs{
+				// With write buffering enabled on the KV client, we need to
+				// ensure that committing a write to the raft log returns only
+				// when it's applied to pebble (since we scan the engine
+				// directly in slurpUserDataKVs).
+				DisableCanAckBeforeApplication: true,
+			},
+		},
+	})
 	defer srv.Stopper().Stop(ctx)
 	s := srv.ApplicationLayer()
 	codec := s.Codec()
@@ -88,9 +98,13 @@ func TestRowFetcherMVCCMetadata(t *testing.T) {
 		FAMILY (a, b, c), FAMILY (d)
 	)`)
 	desc := desctestutils.TestingGetPublicTableDescriptor(kvDB, codec, `d`, `parent`)
+	// Request crdb_internal_mvcc_timestamp column in addition to all user
+	// columns from the table.
+	fetchColumnIDs := desc.PublicColumnIDs()
+	fetchColumnIDs = append(fetchColumnIDs, colinfo.MVCCTimestampColumnID)
 	var spec fetchpb.IndexFetchSpec
 	if err := rowenc.InitIndexFetchSpec(
-		&spec, codec, desc, desc.GetPrimaryIndex(), desc.PublicColumnIDs(),
+		&spec, codec, desc, desc.GetPrimaryIndex(), fetchColumnIDs,
 	); err != nil {
 		t.Fatal(err)
 	}
@@ -129,12 +143,13 @@ func TestRowFetcherMVCCMetadata(t *testing.T) {
 				break
 			}
 			row := rowWithMVCCMetadata{
-				RowIsDeleted:    rf.RowIsDeleted(),
-				RowLastModified: eval.TimestampToDecimalDatum(rf.RowLastModified()).String(),
+				RowIsDeleted: rf.RowIsDeleted(),
 			}
 			for _, datum := range datums {
 				if datum == tree.DNull {
 					row.PrimaryKey = append(row.PrimaryKey, `NULL`)
+				} else if d, ok := datum.(*tree.DDecimal); ok {
+					row.RowLastModified = d.String()
 				} else {
 					row.PrimaryKey = append(row.PrimaryKey, string(*datum.(*tree.DString)))
 				}
@@ -151,8 +166,8 @@ func TestRowFetcherMVCCMetadata(t *testing.T) {
 	END;`).Scan(&ts1)
 
 	if actual, expected := kvsToRows(slurpUserDataKVs(t, store.TODOEngine(), codec)), []rowWithMVCCMetadata{
-		{[]string{`1`, `a`, `a`, `a`}, false, ts1},
-		{[]string{`2`, `b`, `b`, `b`}, false, ts1},
+		{PrimaryKey: []string{`1`, `a`, `a`, `a`}, RowIsDeleted: false, RowLastModified: ts1},
+		{PrimaryKey: []string{`2`, `b`, `b`, `b`}, RowIsDeleted: false, RowLastModified: ts1},
 	}; !reflect.DeepEqual(expected, actual) {
 		t.Errorf(`expected %v got %v`, expected, actual)
 	}
@@ -165,8 +180,8 @@ func TestRowFetcherMVCCMetadata(t *testing.T) {
 	END;`).Scan(&ts2)
 
 	if actual, expected := kvsToRows(slurpUserDataKVs(t, store.TODOEngine(), codec)), []rowWithMVCCMetadata{
-		{[]string{`1`, `NULL`, `NULL`, `NULL`}, false, ts2},
-		{[]string{`2`, `b`, `b`, `NULL`}, false, ts2},
+		{PrimaryKey: []string{`1`, `NULL`, `NULL`, `NULL`}, RowIsDeleted: false, RowLastModified: ts2},
+		{PrimaryKey: []string{`2`, `b`, `b`, `NULL`}, RowIsDeleted: false, RowLastModified: ts2},
 	}; !reflect.DeepEqual(expected, actual) {
 		t.Errorf(`expected %v got %v`, expected, actual)
 	}
@@ -177,8 +192,8 @@ func TestRowFetcherMVCCMetadata(t *testing.T) {
 		SELECT cluster_logical_timestamp();
 	END;`).Scan(&ts3)
 	if actual, expected := kvsToRows(slurpUserDataKVs(t, store.TODOEngine(), codec)), []rowWithMVCCMetadata{
-		{[]string{`1`, `NULL`, `NULL`, `NULL`}, true, ts3},
-		{[]string{`2`, `b`, `b`, `NULL`}, false, ts2},
+		{PrimaryKey: []string{`1`, `NULL`, `NULL`, `NULL`}, RowIsDeleted: true, RowLastModified: ts3},
+		{PrimaryKey: []string{`2`, `b`, `b`, `NULL`}, RowIsDeleted: false, RowLastModified: ts2},
 	}; !reflect.DeepEqual(expected, actual) {
 		t.Errorf(`expected %v got %v`, expected, actual)
 	}

